src/methods/scgpt_fine_tuned/script.py

```diff
@@ -0,0 +1,503 @@
+schedule_interval = 1
+
+DAB_separate_optim = True if training_settings["DAB"] > 1 else False
+
+logger = scgpt.logger
+scgpt.utils.add_file_handler(logger, "run.log")
+
+
+### Load and pre-process data
+
+# Merge datasets to appy the same pre-processing steps
+input_train.obs["celltype"] = input_train.obs["label"].astype("category")
+input_train.obs["is_test"]  = "0"
+input_test.obs["is_test"]  = "1" 
+input_train = ad.concat([input_train, input_test], join="outer", merge="same", uns_merge="same")
+data_is_raw = False
+filter_gene_by_counts = False
+
+# Make the batch category column
+input_train.obs["batch_id"] = input_train.obs["is_test"].astype("category").cat.codes.values
+num_types = len(input_train.obs["celltype"].unique())
+id2type = dict(enumerate(input_train.obs["celltype"].astype("category").cat.categories))
+input_train.obs["celltype_id"] = input_train.obs["celltype"].astype("category").cat.codes.values
+
+from pathlib import Path
+model_config_file = model_dir / "args.json"
+model_file = model_dir / "best_model.pt"
+vocab_file = model_dir / "vocab.json"
+vocab = scgpt.tokenizer.gene_tokenizer.GeneVocab.from_file(vocab_file)
+for token in special_tokens:
+  if token not in vocab:
+    vocab.append_token(token)
+
+input_train.var["id_in_vocab"] = [
+  1 if gene in vocab else -1 for gene in input_train.var["feature_name"]
+]    
+gene_ids_in_vocab = np.array(input_train.var["id_in_vocab"])
+input_train = input_train[:, input_train.var["id_in_vocab"] >= 0]
+
+logger.info(
+  f"match {np.sum(gene_ids_in_vocab >= 0)}/{len(gene_ids_in_vocab)} genes "
+  f"in vocabulary of size {len(vocab)}."
+)
+
+# Pre-trained model
+with open(model_config_file, "r") as f:
+  model_configs = json.load(f)
+logger.info(
+  f"Resume model from {model_file}, the model args will override the "
+  f"config {model_config_file}."
+)
+embsize = model_configs["embsize"]
+nhead = model_configs["nheads"]
+d_hid = model_configs["d_hid"]
+nlayers = model_configs["nlayers"]
+n_layers_cls = model_configs["n_layers_cls"]
+
+preprocessor = scgpt.preprocess.Preprocessor(
+  use_key="counts",  # the key in input_train.layers to use as raw data
+  filter_gene_by_counts=filter_gene_by_counts, 
+  filter_cell_by_counts=False, 
+  normalize_total=1e4,  # whether to normalize the raw data and to what sum
+  result_normed_key="X_normed",  # the key in input_train.layers to store the normalized data
+  log1p=data_is_raw,  # whether to log1p the normalized data
+  result_log1p_key="X_log1p",
+  subset_hvg=False,  # whether to subset the raw data to highly variable genes
+  hvg_flavor="seurat_v3" if data_is_raw else "cell_ranger",
+  binning=hyperparameters["n_bins"],  # whether to bin the raw data and to what number of bins
+  result_binned_key="X_binned",  # the key in input_train.layers to store the binned data
+)
+
+input_test = input_train[input_train.obs["is_test"] == "1"]
+input_train = input_train[input_train.obs["is_test"] == "0"]
```

Comment:
Is there a reason for doing it this way rather than preprocessing each separately? Is this how the tutorial does it?